api/ccapi/option3_hide_updated.h

```diff
@@ -0,0 +1,83 @@
+// Layer.h (ccapi)
+class Layer {
+  public:
+    Layer();
+    virtual foo() = 0;
+    virtual bar() = 0;
+};
+
+// LayerInternal.h (internal)
+class LayerInternal : public Layer {
+  public:
+    LayerInternal();
+    virtual foo() { /** do some work */ }
+
+  private:
+    // Actual data members
+    // weights, etc
+};
+
+// layer_factory.h (ccapi)
+std::unique_ptr<Layer> createLayer(string layer_type, std::vector<string> properties);
+
+// layer_factory.cpp (internal)
+#include <fclayer.h>
+#include <conv2dlayer.h>
+std::unique_ptr<Layer> createLayer(string layer_type, std::vector<string> properties) {
+  // This can be case-insensitive comparison
+  if (layer_type == "FullyConnected") {
+    FCLayer fclayer = FCLayer();
```

Comment:
constructors with `std::vector<string>` can be added but its just syntactic sugar.